thorlcr/graph/thgraphmaster.cpp

```diff
@@ -2529,14 +2529,31 @@ void CMasterGraph::getFinalProgress()
     msg.append(queryGraphId());
     jobM->broadcast(queryNodeComm(), msg, masterSlaveMpTag, LONGTIMEOUT, "graphEnd", NULL, true);
 
+    Owned<IBitSet> respondedBitSet = createBitSet();
     unsigned n=queryJob().queryNodes();
     while (n--)
     {
         rank_t sender;
         if (!queryNodeComm().recv(msg, RANK_ALL, replyTag, &sender, LONGTIMEOUT))
         {
-            GraphPrintLog("Timeout receiving final progress from slaves, %d slaves did not respond", n+1);
-            return;
+            StringBuffer slaveList;
+            n=queryJob().queryNodes();
+            unsigned s = 0;
+            for (;;)
+            {
+                unsigned ns = respondedBitSet->scan(s, true);
+                while (s<ns && s<n) // NB: if scan find nothing else set, ns = (unsigned)-1
+                {
+                    if (slaveList.length())
+                        slaveList.append(",");
+                    ++s; // inc. 1st as slaveList is list of slaves that are 1 based.
+                    slaveList.append(s);
+                }
+                ++s;
+                if (s>=n)
+                    break;
+            }
+            throw MakeGraphException(this, 0, "Timeout receiving final progress from slaves - these slaves failed to respond: %s", slaveList.str());
```

Comment:
It would, but that's intentional, don't want to scan from 3 (which was last scan result), want next scan be from 4.